### ISSUE TITLE
Replacing use of fp.to_ieee_bv with equisatisfiable transform results in strange behaviour

### ISSUE THREAD

**Issue by delcypher** [Z3Prover/z3#970]
In a tool I'm developing I use the `fp.to_ieee_bv` the non standard function (via the `Z3_mk_fpa_to_ieee_bv` C API function) for converting a floating point sort to a bitvector sort.

I'm currently in the process of gathering solver queries from my tool for submission to SMT-COMP. I have discussed this with @wintersteiger and it is desirable to avoid having `fp.to_ieee_bv` in logged queries because it is not part of the SMT-LIBv2.5 standard.

I implemented an option in my tool so I can avoid using `Z3_mk_fpa_to_ieee_bv()` by introducing a new fresh variable of bitvector sort and adding a constraint that converting the fresh variable to the floating point sort is equal to the floating point expression I am trying to convert. This is analogous to the [transform described in the SMT-LIB description of the floating point theory](http://smtlib.cs.uiowa.edu/theories-FloatingPoint.shtml).

When I did this I started running my tool's regression test suite and I hit a problem. I have a test that completes in a very short amount of time when my tool uses `Z3_mk_fpa_to_ieee_bv()`. When I take that same test but use an encoding that avoids `Z3_mk_fpa_to_ieee_bv()`  my tools gets stuck waiting for Z3 to return the satisfiability of a query.

To debug this I first tried logging the queries that were being were issued by my tool (when running each mode: one using `Z3_mk_fpa_to_ieee_bv()` and one not avoids using that function) and replaying them using the `z3` binary. I observed that

* `z3` returned an answer very quickly for **all the queries**.
* For the 9th query (counting from 1) the **satisfiability changes**. What I mean by this is when I run my tool in the two different modes queries up until this point are identical and have the same satisfiability.
In the mode where `Z3_mk_fpa_to_ieee_bv()` is used the query (`isolated_use_fp.to_ieee_bv.smt2` file)  is reported as `unsat` but in the mode that doesn't use `Z3_mk_fpa_to_ieee_bv()` it is reported as `sat` ( `isolated_no_use_fp.to_ieee_bv.smt2`  file).

As far as I can tell the query where the satisfiability changes is the query that my tool seems to get stuck on (despite the query being handled promptly when I give it to `z3` as an SMT-LIBv2.5 query).

I also had my tool record the Z3 interaction log, once using the `Z3_mk_fpa_to_ieee_bv()` encoding ( see `z3.use_fp.to_ieee_bv.intlog` file) and once using the other encoding (see `z3.no_use_fp.to_ieee_bv.intlog` file.). When I replay the interaction log I can reproduce the behaviour in my tool where using the encoding that avoids `Z3_mk_fpa_to_ieee_bv()` seems to cause Z3 to spend a long time trying to determine the satisfiability.

**Side note: When I pass `-v:20` when replaying the log I see lots of `(smt ...)` messages which I presume means it is using the `smt` tactic. This kind of surprises me. The logic being used in these queries is `QF_FPBV` which I would have thought would use a tactic that bit blasts to a SAT solver.**

I've attached a zip file containing the following files to help debug this.

* `isolated_no_use_fp.to_ieee_bv.smt2` - The query where satisifiability changes. This is the encoding that avoids using `fp.to_ieee_bv`.
* `isolated_use_fp.to_ieee_bv.smt2`  - The query where satisifiability changes.  This is the encoding that uses `fp.to_ieee_bv`.
* `use.fp_to_ieee_bv.smt2` - Sequence of recorded queries from my tool. This is the encoding that uses `fp.to_ieee_bv`.
* `no_use.fp_to_ieee_bv.smt2` - Sequence of recorded queries from my tool. This is the encoding that avoids using `fp.to_ieee_bv`.
* `z3.no_use_fp.to_ieee_bv.intlog` - Z3 interaction log from my tool. This uses the encoding that avoids using `fp.to_ieee_bv`.
* `z3.use_fp.to_ieee_bv.intlog` - Z3 interaction log from my tool. This uses the encoding that uses `fp.to_ieee_bv`.

For your reference the version of Z3 that was used here is 80c10d58333f85184608d3e4a98f6544505f8398

So it seems there are two issues here.

* When I use an encoding that avoids  `Z3_mk_fpa_to_ieee_bv()` satisfiability seems to change.
* I observe Z3 getting stuck on a query that if I execute outside of my tool Z3 solves very quickly.

Any ideas on what is going wrong here?

# Changing satisfiability

**TL;DR: Only read this if you want to understand the encoding used in the constraints**

I think this might be worth discussing because I might have screwed up here. It might also help a reader under the queries being issued.

Although I have tried to avoid using `Z3_mk_fpa_to_ieee_bv()` correctly I might have made a mistake.

The query that is causing a problem is modelling a C program that uses `long double` type on `x86_64` hardware. On this machine this type is an 80-bit wide float that does not use an IEEE-754 binary encoding. This type has a binary encoding that it is very similar to an IEEE-754 encoding except that the integer portion of the significand is **explicit**.

There is no sort in Z3 that represents this however we attempt to model this in our tool as follows.

* When we want to convert `(_ BitVec 80)` to a floating point sort we create an expression where explicit integer bit in the significand is removed (i.e. a bitvector that 79-bits wide) so we have an IEEE-754 binary encoding and then convert this to a `(FloatingPoint 15 64)` sort. To be sound when we do this conversion we add an extra constraint on the explicit integer bit.
```
(declare-fun x_ackermann_0 () (_ BitVec 80))
(declare-fun fresh_to_ieee_bv_!0 () (_ BitVec 80))

; Example conversion. (_ BitVec 80) to (FloatingPoint 15 64)
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0)))))
  (not (fp.isNaN a!1))))

; Additional side constraint required that constrains the explicit
; integer significand bit to have the correct value.
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0)))))
  (= ((_ extract 63 63) x_ackermann_0)
     (ite (or (fp.isSubnormal a!1) (fp.isZero a!1)) #b0 #b1))))
```

* When we want to convert from `(FloatingPoint 15 64)` to `(_ BitVec 80)` we convert to IEEE bits (using `Z3_mk_fpa_to_ieee_bv()` or using the other encoding) and then extract the sign bit, exponent bit and significand and concatenate them with the explicit integer bit inserted (this integer bit is `a!2` in the example below). This conversion is `a!4` in the example below.

```
(declare-fun x_ackermann_0 () (_ BitVec 80))
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0)))))
(let ((a!2 (concat ((_ extract 78 78)
                     (fp.to_ieee_bv (fp.sqrt roundNearestTiesToEven a!1)))
                   ((_ extract 77 63)
                     (fp.to_ieee_bv (fp.sqrt roundNearestTiesToEven a!1)))))
      (a!3 (ite (or (fp.isSubnormal (fp.sqrt roundNearestTiesToEven a!1))
                    (fp.isZero (fp.sqrt roundNearestTiesToEven a!1)))
                #b0
                #b1)))
(let ((a!4 (concat (concat a!2 a!3)
                   ((_ extract 62 0)
                     (fp.to_ieee_bv (fp.sqrt roundNearestTiesToEven a!1))))))
  (bvsle #x00000000000000000000 a!4)))))

; Side constraint for (_ BitVec 80) to `(FloatingPoint 15 64)`  conversion
;
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0)))))
  (= ((_ extract 63 63) x_ackermann_0)
     (ite (or (fp.isSubnormal a!1) (fp.isZero a!1)) #b0 #b1))))

```

This complicated conversion might be interacting with my replacement for `Z3_mk_fpa_to_ieee_bv()` in some bad way that I have not foreseen. The encoding of the above without using `Z3_mk_fpa_to_ieee_bv()`  looks like this

```
(declare-fun x_ackermann_0 () (_ BitVec 80))
(declare-fun fresh_to_ieee_bv_!0 () (_ BitVec 80))

(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0)))))
(let ((a!2 (ite (or (fp.isSubnormal (fp.sqrt roundNearestTiesToEven a!1))
                    (fp.isZero (fp.sqrt roundNearestTiesToEven a!1)))
                #b0
                #b1)))
(let ((a!3 (concat (concat (concat ((_ extract 78 78) fresh_to_ieee_bv_!0)
                                   ((_ extract 77 63) fresh_to_ieee_bv_!0))
                           a!2)
                   ((_ extract 62 0) fresh_to_ieee_bv_!0))))
  (bvsle #x00000000000000000000 a!3)))))

; Side constraint for `(FloatingPoint 15 64)` to `(_ BitVec 80)` conversion
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0))))
      (a!2 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) fresh_to_ieee_bv_!0)
                             ((_ extract 78 64) fresh_to_ieee_bv_!0))
                     ((_ extract 62 0) fresh_to_ieee_bv_!0)))))
  (= (fp.sqrt roundNearestTiesToEven a!1) a!2)))


; Side constraint for `(_ BitVec 80)` to `(FloatingPoint 15 64)`  conversion
;
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) x_ackermann_0)
                             ((_ extract 78 64) x_ackermann_0))
                     ((_ extract 62 0) x_ackermann_0)))))
  (= ((_ extract 63 63) x_ackermann_0)
     (ite (or (fp.isSubnormal a!1) (fp.isZero a!1)) #b0 #b1))))_0)

; Side constraint for `(FloatingPoint 15 64)` to `(_ BitVec 80)` conversion.
; The encoding that avoids fp.to_ieee_bv does the conversion the other way
; so we get this constraint that makes sure the integer significand bit of the fresh
; variable has the correct value.
(assert (let ((a!1 ((_ to_fp 15 64)
             (concat (concat ((_ extract 79 79) fresh_to_ieee_bv_!0)
                             ((_ extract 78 64) fresh_to_ieee_bv_!0))
                     ((_ extract 62 0) fresh_to_ieee_bv_!0)))))
  (= ((_ extract 63 63) fresh_to_ieee_bv_!0)
     (ite (or (fp.isSubnormal a!1) (fp.isZero a!1)) #b0 #b1))))

```

I think my encoding is right but maybe I made a mistake somewhere?

**Comment by delcypher** [comment]
Here are the interaction logs and queries as discussed above.
[int_log_and_queries.zip](https://github.com/Z3Prover/z3/files/908345/int_log_and_queries.zip)

**Comment by wintersteiger** [comment]
Regarding performance: Yes, if uninterpreted functions are used, Z3 will use the SMT solver and not the SAT solver, because the latter doesn't support UFs. The [internal encoding of to_ieee_bv](https://github.com/Z3Prover/z3/blob/master/src/ast/fpa/fpa2bv_converter.cpp#L3063) only uses UFs for NaNs when `m_hi_fp_unspecified` (command line option `rewriter.hi_fp_unspecified`) is enabled (default is off). At this point I don't know whether we can get the same behavior by only using SMT-standard concepts, as that doesn't give us access to individual bits of the encoding (because they assume that (at least some) SMT solvers won't use an encoding based on bit-vectors where this is a no-op).

**Comment by wintersteiger** [comment]
Regarding soundness of those 80-bit floats: Yes, likely there's a problem with this somewhere. Obviously, NaNs are always special, so check those first :-)

My immediate hunch is that sub-normal numbers won't work correctly, because the integer 0-bit is in the wrong place (may need re-normalization). The rest is TL;DR indeed!

**Comment by delcypher** [comment]
@wintersteiger 
> Regarding soundness of those 80-bit floats: Yes, likely there's a problem with this somewhere. Obviously, NaNs are always special, so check those first :-)
>
> My immediate hunch is that sub-normal numbers won't work correctly, because the integer 0-bit is in the wrong place (may need re-normalization). The rest is TL;DR indeed!

I've spent more time looking at this and I found a bug in the way I was encoding long doubles. I fixed it and now the problem of an "unsat" query becoming "sat" is now gone :)

So now the only remaining problem is the performance issue I observed. Even though I was asking the wrong query it would be good to understand why interacting with Z3 with the solver API gave very bad performance but doing so via an SMT-LIBv2 file was very fast.

**Comment by delcypher** [comment]
> Regarding performance: Yes, if uninterpreted functions are used, Z3 will use the SMT solver and not the SAT solver, because the latter doesn't support UFs. The internal encoding of to_ieee_bv only uses UFs for NaNs when m_hi_fp_unspecified (command line option rewriter.hi_fp_unspecified) is enabled (default is off).

Thanks for explaining this. I still don't understand why I saw performance problems when issuing queries to Z3 via the C API. Replaying the SMTLIBv2 queries is very fast but replaying the interaction log results in Z3 getting stuck on one of the queries. It's getting stuck when I **don't use** fp.to_ieee_bv so there are no uninterpreted functions.

> At this point I don't know whether we can get the same behavior by only using SMT-standard concepts, as that doesn't give us access to individual bits of the encoding (because they assume that (at least some) SMT solvers won't use an encoding based on bit-vectors where this is a no-op).

Yes when I disabled use of `fp.to_ieee_bv` in my tool several of my tests broke. I had several tests (C programs) that examine the sign bit of a NaN. My tests were passing by accident when using `fp.to_ieee_bv` because I had `rewriter.hi_fp_unspecified` enabled and so I was got the same bit pattern back for a NaN that had the sign bit set to `0`.

Admittedly my tests weren't that well written before and this problem isn't really Z3's fault. I consider these problems a fault in the design of the FloatingPoint theory in that there is only one type of NaN. This makes it very difficult to model the sign bit of a NaN and also quiet/signalling NaNs.

**Comment by wintersteiger** [comment]
Yes, there are definitely some issues with the SMT2 FP theory. At the time the reasoning was to make it easier for solver developers to drive adoption. The single NaN does make it harder on the modelling side though. 

Could you send me your input and interaction log for the query that behaves differently when replayed? The difference is likely due to different parameters, either involuntarily induced on your end, or because the parameters aren't passed on correctly through the tactic framework (we fixed bunch of those, but I'm sure some still remain).

**Comment by delcypher** [comment]
@wintersteiger 
> Could you send me your input and interaction log for the query that behaves differently when replayed? The difference is likely due to different parameters, either involuntarily induced on your end, or because the parameters aren't passed on correctly through the tactic framework (we fixed bunch of those, but I'm sure some still remain).

They are included in [this zip file](https://github.com/Z3Prover/z3/files/908345/int_log_and_queries.zip).

The interaction log is the `z3.no_use_fp.to_ieee_bv.intlog` file and corresponding SMTLIB queries are `no_use.fp_to_ieee_bv.smt2`.

I'm using `Z3_mk_simple_solver()` to make the solver so I am not explicitly using any particular tactic, is that the problem? I don't use `Z3_mk_solver_for_logic()` because that is not a constant (it is either `QF_BV`, `QF_ABV`, `QF_FPBV` or `QF_FPABV`). Should I be using Z3's probe and tactic API instead to build a custom solver that applies a particular tactic based on the results of the probes?

**Comment by wintersteiger** [comment]
Sorry, my latency on this one was too high, now the logs don't replay anymore. Generally though, yes, using the simple solver may have been the problem. In that case, Z3 will use the SMT solver without any of the pre-processing tactics (but still with it's own internal pre-processing) or it's logic detection probes, thus performance can differ wildly. For instance, the `qffp` tactic (for QF_FP problems) eagerly translates all floats into bit-vectors and it tries to ackermannize some of the UFs. The same problems being fed into the simple solver (or the `smt` tactic) will translate the floats lazily, when the relevancy heuristics think they matter, and it will keep all UFs. Obviously, that can make a huge difference in performance.

**Comment by wintersteiger** [comment]
Still no repro, so I'm closing this issue. Please open a new issue if this comes up again. For anyone reading this: `fp.to_ieee_bv` should be avoided wherever possible. 

The SMT FP theory still supports only one NaN and so does Z3. So, an application would have to wrap the `fp.*` calls and replace the NaNs in the results. It's not impossible to add that to Z3, but it would be significant effort and so far I have had only two requests, both non-blocking for the respective applications.
